Terraform sync tool (#290)
GoogleCloudPlatform/bigquery-utils: tools/terraform_sync_tool/README.md

# Terraform Sync Tool

> This directory contains the setup for the Terraform Sync Tool. Terraform Sync Tool was designed to address the schema drifts in BigQuery tables and keep the 
> Terraform schemas up-to-date with the BigQuery table schemas in production environment. Schema drifts occurred when BigQuery Table schemas are updated by newly 
> ingested data while Terraform schema files contain the outdated schemas. Therefore, this tool will detect the schema drifts, trace the origins of the drifts, and alert
> developers/data engineers.

Author: ```suggestion
This directory contains the Terraform Sync Tool. This tool intentionally fails your CI/CD pipeline when schema drifts occur between what your BigQuery Terraform resources declare and what's actually present in your BigQuery environment. Theses schema drifts happen when BigQuery tables are updated by processes outside of Terraform (ETL process may dynamically add new columns when loading data into BigQuery). When drifts occur, you end up with outdated BigQuery Terraform resource files. This tool detects the schema drifts, traces the origins of the drifts, and alerts developers/data engineers (by failing the CI/CD pipeline) so they can patch the Terraform in their current commit.
```

Reviewer: updated in https://github.com/GoogleCloudPlatform/bigquery-utils/pull/290/commits/7f469e47bf48003dfc90c9fa406398d1ae927ad8

> The Terraform Schema Sync Tool fails the build attemps if resource drifts are detected and notifies the latest resource information. Developers and data engineers should be able to update the Terraform resources accordingly. 

Author: ```suggestion
```

Reviewer: Updated in https://github.com/GoogleCloudPlatform/bigquery-utils/pull/290/commits/7f469e47bf48003dfc90c9fa406398d1ae927ad8

Terraform Sync Tool can be integrated into your CI/CD pipeline. You'll need to add two steps to CI/CD pipeline. 

> - Step 0: Use Terraform/erragrunt command to detect resource drifts and write output into a JSON file

Author: ```suggestion
- Step 0: Run the Terraform plan command (using either Terraform/Terragrunt) with the `-json` option  and write the output into a JSON file using the caret operator `> output.json`
```

- Step 1: Use Python scripts to identify and investigate the drifts

> ## How to run Terraform Schema Sync Tool

Reviewer: Updated in https://github.com/GoogleCloudPlatform/bigquery-utils/pull/290/commits/5496102bcbf327ed267d8b018d2e8d599d05f84f and https://github.com/GoogleCloudPlatform/bigquery-utils/pull/290/commits/34f14611219b4bb73f3ff9bb7a892af9e97a2d5f

> #### Use Terraform/Terragrunt commands to test if any resources drifts existed

Author: All these should be 3rd level headers (###), not 4 (####)

Terragrunt/Terraform commands:
```

> terragrunt run-all plan -json --terragrunt-non-interactive

Author: ```suggestion
terragrunt run-all plan -json --terragrunt-non-interactive > plan_output.json
```

# Terraform Command
terraform plan -json
```

> After running the Terrform plan command, **the event type "resource_drift"("type": "resource_drift") indicates a drift has occurred**.
> If drifts detected, please update your terraform configurations and address the resource drifts based on the event outputs.
> 
> 
> #### Add Could Build Steps to your configuration file
> 
> Please check cloud build steps in `cloudbuild.yaml` file, and add these steps to your Cloud Build Configuration File.
> 
> - step 0: run terraform commands in `deploy.sh` to detects drifts
> 
> Add `deploy.sh` to your project directory. 
> 
> - step 1: run python scripts to investigate terraform output
> 
> Add `requirements.txt` and `terraform_sync.py` to your project directory.
> 
> #### (Optional if you haven't created Cloud Build Trigger) Create and configure a new Trigger in Cloud Build
> Make sure to indicate your cloud configuration file location correctly.
> 
> #### That's all you need! Let's commit and test in CLoud Build!

Author: ```suggestion
```

> Terragrunt/Terraform commands:
> ```
> terragrunt run-all plan -json --terragrunt-non-interactive
> 
> # Terraform Command
> terraform plan -json
> ```
> 
> After running the Terrform plan command, **the event type "resource_drift"("type": "resource_drift") indicates a drift has occurred**.
> If drifts detected, please update your terraform configurations and address the resource drifts based on the event outputs.
> 
> 
> #### Add Could Build Steps to your configuration file
> 
> Please check cloud build steps in `cloudbuild.yaml` file, and add these steps to your Cloud Build Configuration File.
> 
> - step 0: run terraform commands in `deploy.sh` to detects drifts
> 
> Add `deploy.sh` to your project directory. 
> 
> - step 1: run python scripts to investigate terraform output
> 
> Add `requirements.txt` and `terraform_sync.py` to your project directory.
> 
> #### (Optional if you haven't created Cloud Build Trigger) Create and configure a new Trigger in Cloud Build
> Make sure to indicate your cloud configuration file location correctly.
> 
> #### That's all you need! Let's commit and test in CLoud Build!
> 

Author: ```suggestion
```

Author: These are not required to run this tool. This is more related to running the example you provide with this tool.

Reviewer: Updated in https://github.com/GoogleCloudPlatform/bigquery-utils/pull/290/commits/5496102bcbf327ed267d8b018d2e8d599d05f84f

To interpret the message, the expected table schema is in the format of [{table1_id:table1_schema}, {table2_id: table2_schema}, ...... ]. table_id falls in the format of [gcp_project_id].[dataset_id].[table_id] 

#### Folder Structure ####
This directory serves as a starting point for your cloud project with terraform-sync-tool as one of qa tools integrated.

    .
    ├── modules                     # Terraform modules directory
    │   ├── bigquery                # Example Terraform BigQuery Setup
    │   └── ...                     # Other modules setup you have
    ├── qa                          # qa environment directory
    │   ├── terragrunt.hcl      
    │   └── terraform-sync-tool     # Tool terraform-sync-tool
    │           ├── json_schemas    # Terraform schema files 
    │           ├── terragrunt.hcl
    │           └── ...
    ├── cloudbuild.yaml             # Cloud Build configuration file

>     ├── deploy.sh                   # Build Step 0 - contains terragrunt commands

Author: All of these should go inside an example/ directory

Reviewer: Updated in https://github.com/GoogleCloudPlatform/bigquery-utils/pull/290/commits/5496102bcbf327ed267d8b018d2e8d599d05f84f

# Terraform Sync Tool

This directory contains the setup for the Terraform Sync Tool. Terraform Sync Tool was designed to address the schema drifts in BigQuery tables and keep the 
Terraform schemas up-to-date with the BigQuery table schemas in production environment.

Terraform Sync Tool can be integrated into your CI/CD pipeline using Cloud Build. You'll need to add two steps to `cloudbuild.yaml`. 
- Step 0: Use Terragrunt command to detect resource drifts and write output into a JSON file
- Step 1: Use Python scripts to identify and investigate the drifts

Cloud Build fails the build attemps if resource drifts are detected and notifies the latest resource information. Developers should be able to update
the Terraform resources accordingly. 

## Prerequisite
Before building the terraform sync tool, please ensure that billing and Cloud Build are enabled for your Cloud project.

> You'll also need to install Cloud SDK(https://cloud.google.com/sdk/docs/install) and Terragrunt(https://terragrunt.gruntwork.io/docs/getting-started/install)

Author: Cloud SDK is not a necessary requirement

Reviewer: README Updated

    │   └── ...                     # Other modules setup you have
    ├── qa                          # qa environment directory
    │   ├── terragrunt.hcl      
    │   └── terraform-sync-tool     # Tool terraform-sync-tool
    │           ├── json_schemas    # Terraform schema files 
    │           ├── terragrunt.hcl
    │           └── ...
    ├── cloudbuild.yaml             # Cloud Build configuration file
    ├── deploy.sh                   # Build Step 0 - contains terragrunt commands
    ├── requirements.txt            # Build Step 1 - Specifies python dependencies
    ├── terraform_sync.py           # Build Step 1 - python scripts
    └── ...                         # etc.

## TODO
Please make sure to update 
- **YOUR_GCP_PROJECT_ID** in `./qa/terragrunt.hcl` 
- **YOUR_BUCKET_NAME** in `./qa/terragrunt.hcl` 
- **YOUR_DATASET_ID** in `./qa/terraform-sync-tool/terragrunt.hcl` 

Please **create and configure your trigger in Cloud Build** and make sure to use `cloudbuild.yaml` as **Cloud Build configuration file location**

## Setup

Use Cloud SDK to set the specified property in your active configuration only
```

> gcloud config set project <your-project-id>

Author: Either derive project id from Application Default Credentials or make it an input argument to the python tool
